storage-proofs/src/circuit/stacked/hash.rs

```diff
@@ -0,0 +1,147 @@
+where
+    E: JubjubEngine,
+    CS: ConstraintSystem<E>,
+{
+    let mut values = Vec::new();
+    values.extend_from_slice(first);
+
+    // pad to full bytes
+    while values.len() % 8 > 0 {
+        values.push(Boolean::Constant(false));
+    }
+
+    values.extend_from_slice(second);
+    // pad to full bytes
+    while values.len() % 8 > 0 {
+        values.push(Boolean::Constant(false));
+    }
+
+    hash1(cs.namespace(|| "hash2"), params, &values)
+}
+
+/// Hash a list of bits.
+pub fn hash1<E, CS>(
+    mut cs: CS,
+    params: &E::Params,
+    values: &[Boolean],
+) -> Result<num::AllocatedNum<E>, SynthesisError>
+where
+    E: JubjubEngine,
+    CS: ConstraintSystem<E>,
+{
+    assert!(values.len() % 32 == 0, "input must be a multiple of 32bits");
+
+    if values.is_empty() {
+        // can happen with small layers
+        num::AllocatedNum::alloc(cs.namespace(|| "hash1"), || Ok(E::Fr::zero()))
+    } else if values.len() > PEDERSEN_BLOCK_SIZE {
+        pedersen_md_no_padding(cs.namespace(|| "hash1"), params, values)
+    } else {
+        pedersen(cs.namespace(|| "hash1"), params, values)
+    }
+}
+
+/// Hash a list of bits.
+pub fn hash_single_column<E, CS>(
+    mut cs: CS,
+    params: &E::Params,
+    rows: &[Option<E::Fr>],
+) -> Result<num::AllocatedNum<E>, SynthesisError>
+where
+    E: JubjubEngine,
+    CS: ConstraintSystem<E>,
+{
+    let mut bits = Vec::new();
+    for (i, row) in rows.iter().enumerate() {
+        let row_num = num::AllocatedNum::alloc(
+            cs.namespace(|| format!("hash_single_column_row_{}_num", i)),
+            || {
+                row.map(Into::into)
+                    .ok_or_else(|| SynthesisError::AssignmentMissing)
+            },
+        )?;
+        let mut row_bits =
+            row_num.into_bits_le(cs.namespace(|| format!("hash_single_column_row_{}_bits", i)))?;
+        // pad to full bytes
+        while row_bits.len() % 8 > 0 {
+            row_bits.push(Boolean::Constant(false));
+        }
+        bits.extend(row_bits);
+    }
+
+    hash1(cs.namespace(|| "hash_single_column"), params, &bits)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use bellperson::ConstraintSystem;
+    use fil_sapling_crypto::circuit::boolean::Boolean;
+    use fil_sapling_crypto::jubjub::JubjubBls12;
+    use paired::bls12_381::{Bls12, Fr};
+    use rand::{Rng, SeedableRng, XorShiftRng};
+
+    use crate::circuit::test::TestConstraintSystem;
+    use crate::fr32::fr_into_bytes;
+    use crate::settings;
+    use crate::stacked::hash::hash2 as vanilla_hash2;
+    use crate::util::bytes_into_boolean_vec;
+
+    #[test]
+    fn test_hash2_circuit() {
+        let mut rng = XorShiftRng::from_seed([0x5dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
+        let window_size = settings::SETTINGS
+            .lock()
+            .unwrap()
+            .pedersen_hash_exp_window_size;
+        let params = &JubjubBls12::new_with_window_size(window_size);
```

Comment:
We actually already create the Pedersen Hash params here: https://github.com/filecoin-project/rust-fil-proofs/blob/feat/stacked/storage-proofs/src/crypto/pedersen.rs#L11

I know that there are many places in the code where we are recreating this value even though we have it statically allocated. Just a heads up because the Pedersen Hash params take 1 or 2 seconds to generate.

Reply:
yes, created https://github.com/filecoin-project/rust-fil-proofs/issues/906 to do this for the whole codebase